embassy-extras/src/peripheral.rs

```diff
@@ -2,14 +2,46 @@ use core::cell::UnsafeCell;
 use core::marker::{PhantomData, PhantomPinned};
 use core::pin::Pin;
 
+use cortex_m::peripheral::scb::{Exception, SystemHandler, VectActive};
+use cortex_m::peripheral::{NVIC, SCB};
 use embassy::interrupt::{Interrupt, InterruptExt};
 
-pub trait PeripheralState {
+/// A version of `PeripheralState` without the `'static` bound,
+/// for cases where the compiler can't statically make sure
+/// that `on_interrupt` doesn't reference anything which might be invalidated.
+///
+/// # Safety
+/// When types implementing this trait are used with `PeripheralMutex`,
+/// no fields referenced by `on_interrupt`'s lifetimes must end without first calling `Drop` on the `PeripheralMutex`.
+pub unsafe trait PeripheralStateUnchecked: Send {
     type Interrupt: Interrupt;
     fn on_interrupt(&mut self);
 }
 
-pub struct PeripheralMutex<S: PeripheralState> {
+/// A type which can be used as state with `PeripheralMutex`.
+///
+/// It needs to be `Send` because `&mut` references are sent back and forth between the 'thread' which owns the `PeripheralMutex` and the interrupt,
+/// and `&mut T` is `Send` where `T: Send`.
+///
+/// It also requires `'static`, because although `Pin` guarantees that the memory of the state won't be invalidated,
+/// it doesn't guarantee that the lifetime will last.
+pub trait PeripheralState: Send + 'static {
```

Comment:
It'd still be nice to have a have a guaranteed-safe way to share state with an interrupt, so maybe the `'static` bound could be moved to `new`, and there could be an unsafe `new_unchecked` which doesn't have the `'static` bound. That would eliminate the two traits.

Reply:
Ooh, sure! that would work too. Yeah, it's nice having a fully safe API. Having 2 traits was too much complexity for that, but the `new_unchecked` is much simpler, so why not.

Reply:
I ended up making a `register_interrupt_unchecked` method instead, since it's safe for a `PeripheralMutex` with non-`'static` state to exist, just not to enable it.